test/storage-luatest/garbage_collector_test.lua

```diff
@@ -474,6 +504,7 @@ test_group.test_replica_ref_bucket_wait_lsn = function(g)
     g.replica_1_b:exec(function()
         rawset(_G, 'old_replication', box.cfg.replication)
         box.cfg{replication = {}}
+        ivshard.storage.internal.is_bucket_protected = false
```

Comment:
You probably didn't want long enough. I applied this diff and all worked fine:
<details><summary>Diff</summary>

```Diff
diff --git a/test/storage-luatest/garbage_collector_test.lua b/test/storage-luatest/garbage_collector_test.lua
index dd1e311..64221cd 100644
--- a/test/storage-luatest/garbage_collector_test.lua
+++ b/test/storage-luatest/garbage_collector_test.lua
@@ -502,9 +502,6 @@ test_group.test_replica_ref_bucket_wait_lsn = function(g)
     -- Break replication.
     --
     g.replica_1_b:exec(function()
-        -- Disable the protection because will now receive active->sent
-        -- transition from the master.
-        ivshard.storage.internal.is_bucket_protected = false
         rawset(_G, 'old_replication', box.cfg.replication)
         box.cfg{replication = {}}
     end)
@@ -534,6 +531,9 @@ test_group.test_replica_ref_bucket_wait_lsn = function(g)
     -- Restore the replication.
     --
     g.replica_1_b:exec(function()
+        -- Disable the protection because will now receive active->sent
+        -- transition from the master.
+        ivshard.storage.internal.is_bucket_protected = false
         box.cfg{replication = _G.old_replication}
         _G.old_replication = nil
     end)
@@ -783,8 +783,8 @@ test_group.test_unit_gc_bucket_drop = function(g)
         -- Collect second time and nothing happens.
         --
         route_map = {}
-        ivshard.storage.internal.is_bucket_protected = false
         ilt.assert(gc_bucket_drop(bstatus.GARBAGE, route_map))
+        ivshard.storage.internal.is_bucket_protected = false
         ilt.assert(gc_bucket_drop(bstatus.SENT, route_map))
         ivshard.storage.internal.is_bucket_protected = true
         ilt.assert_equals(_bucket:count(), bucket_count)
@@ -974,16 +974,16 @@ test_group.test_unit_bucket_test_gc = function(g)
         --
         -- Cleanup.
         --
-        ivshard.storage.internal.is_bucket_protected = false
         box.begin()
         for _, bid in ipairs(sent_ro_bids) do
             ilt.assert(ivshard.storage.bucket_unrefro(bid))
         end
+        ivshard.storage.internal.is_bucket_protected = false
         for _, bid in ipairs(test_bids) do
             _bucket:update({bid}, {{'=', 2, ivconst.BUCKET.ACTIVE}})
+        ivshard.storage.internal.is_bucket_protected = true
         end
         box.commit()
-        ivshard.storage.internal.is_bucket_protected = true
         --
         -- Long bucket list iteration should yield.
         --
```
</details>
(In the same diff I reduced the unsafe sections a little.)

It can't hang, because `replica_1_b` just turned the replication off in the beginning. It couldn't receive anything until the replication is re-enabled.

Reply:
Yeah, you're right, it can't hang for that reason, overlooked disabling the replication. However, after applying such diff I got the following output:
```text
[001] not ok 30	bucket_gc.engine:"vinyl".test_unit_bucket_test_gc
[001] #   Transaction is active at return from function
[001] #   stack traceback:
[001] #   	builtin/box/net_box.lua:817: in function '_request'
[001] #   	builtin/box/net_box.lua:878: in function 'eval'
[001] #   	...k/vshard/test/storage-luatest/garbage_collector_test.lua:910: in function 'bucket_gc.engine:"vinyl".test_unit_bucket_test_gc'
[001] #   	...
[001] #   	[C]: in function 'xpcall'
[001] # Starting group: bucket_gc.engine:"vinyl".memtx_use_mvcc_engine:true
```

I suppose you meant this (this is diff with your diff):
```diff
diff --git a/test/storage-luatest/garbage_collector_test.lua b/test/storage-luatest/garbage_collector_test.lua
index 64221cd..77760c3 100644
--- a/test/storage-luatest/garbage_collector_test.lua
+++ b/test/storage-luatest/garbage_collector_test.lua
@@ -981,8 +981,8 @@ test_group.test_unit_bucket_test_gc = function(g)
         ivshard.storage.internal.is_bucket_protected = false
         for _, bid in ipairs(test_bids) do
             _bucket:update({bid}, {{'=', 2, ivconst.BUCKET.ACTIVE}})
-        ivshard.storage.internal.is_bucket_protected = true
         end
+        ivshard.storage.internal.is_bucket_protected = true
         box.commit()
         --
         -- Long bucket list iteration should yield.

```